iceoryx_hoofs/include/iceoryx_hoofs/cxx/algorithm.hpp

```diff
@@ -116,6 +118,108 @@ template <typename T, typename... ValueList>
 inline constexpr bool
 doesContainValue(const T value, const T firstValueListEntry, const ValueList... remainingValueListEntries) noexcept;
 } // namespace algorithm
+
+namespace internal
+{
+/// @brief struct to find the best fitting unsigned integer type
+template <bool GreaterUint8, bool GreaterUint16, bool GreaterUint32>
+struct BestFittingTypeImpl
+{
+    using Type_t = uint64_t;
+};
+
+template <>
+struct BestFittingTypeImpl<false, false, false>
+{
+    using Type_t = uint8_t;
+};
+
+template <>
+struct BestFittingTypeImpl<true, false, false>
+{
+    using Type_t = uint16_t;
+};
+
+template <>
+struct BestFittingTypeImpl<true, true, false>
+{
+    using Type_t = uint32_t;
+};
+} // namespace internal
+/// @brief get the best fitting unsigned integer type for a given value at compile time
+template <uint64_t Value>
+struct BestFittingType
+{
+/// ignore the warnings because we need the comparisons to find the best fitting type
+#pragma GCC diagnostic push
+#pragma GCC diagnostic ignored "-Wtype-limits"
+    using Type_t = typename internal::BestFittingTypeImpl<(Value > std::numeric_limits<uint8_t>::max()),
+                                                          (Value > std::numeric_limits<uint16_t>::max()),
+                                                          (Value > std::numeric_limits<uint32_t>::max())>::Type_t;
+#pragma GCC diagnostic pop
+};
+
+template <uint64_t Value>
+using BestFittingType_t = typename BestFittingType<Value>::Type_t;
+
+template <typename T, T Minimum>
+struct greater_or_equal
+{
+  public:
+    // AXIVION Next Construct AutosarC++19_03-A12.1.4: this class should behave like a T but which never can be less
+    // than Minimum. Adding explicit would defeat the purpose.
+    // NOLINTNEXTLINE(hicpp-explicit-conversions)
+    greater_or_equal(T t) noexcept
+        : m_value(t)
+    {
+        cxx::Expects(t >= Minimum);
+    }
```

Comment:
Nothing for this PR but all of these constructs should have a factory method like `optional<Self> Self::from(...)` and not terminate by default. For the performance people there could be a `Self Self::from_unsafe_unchecked(...)`.

I'm also thinking whether `algorithm.hpp` is the right location in the long run. They could also went into the `vocabulary` module as separate files for each class/struct.

Reply:
I think `algorithm.hpp` is the right place in the long run, but let's move this discussion to a follow-up PR.

Reply:
The reason I think this might be the wrong place is that `greater_or_equal` and `range` are types and therefore I don't see them in `algorithm`. They are used to communicate some contracts for the user and in the long run I see these types as a specialization of a `NewType`.

Reply:
@elBoberido While they are indeed types, they calculate something and are hence `math` constructs. Hence, `algorithm` is the closest header for me.

Reply:
@mossmaurice they do not really calculate but ensure the invariant holds on construction or terminate. With this argumentation the `not_null` type must also be part of `algorithm` since it ensures the invariant of not constructing with a `nullptr` ;)

Well maybe just a different way to view at the problem.

Reply:
Yeah, you are right. Considering that `not_null` shall eventually be replaced by references, we could consider moving it to `iceoryx_dust`. However, that's currently not possible because it's used in `loffli`, which resides in `iceoryx_hoofs`. Something for the next PR ;)

Reply:
Sure. I even proposed to skip the moving out of helplets in a follow up PR :)